plume-models/src/lists.rs

```diff
@@ -0,0 +1,588 @@
+use diesel::{self, ExpressionMethods, QueryDsl, RunQueryDsl};
+
+use blogs::Blog;
+use schema::{blogs, list_elems, lists, users};
+use std::convert::{TryFrom, TryInto};
+use users::User;
+use {Connection, Error, Result};
+
+/// Represent what a list is supposed to store. Represented in database as an integer
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ListType {
+    User,
+    Blog,
+    Word,
+    Prefix,
+}
+
+impl TryFrom<i32> for ListType {
+    type Error = ();
+
+    fn try_from(i: i32) -> std::result::Result<Self, ()> {
+        match i {
+            0 => Ok(ListType::User),
+            1 => Ok(ListType::Blog),
+            2 => Ok(ListType::Word),
+            3 => Ok(ListType::Prefix),
+            _ => Err(()),
+        }
+    }
+}
+
+impl Into<i32> for ListType {
+    fn into(self) -> i32 {
+        match self {
+            ListType::User => 0,
+            ListType::Blog => 1,
+            ListType::Word => 2,
+            ListType::Prefix => 3,
+        }
+    }
+}
+
+#[derive(Clone, Queryable, Identifiable)]
+pub struct List {
+    pub id: i32,
+    pub name: String,
+    pub user_id: Option<i32>,
+    type_: i32,
+}
+
+#[derive(Default, Insertable)]
+#[table_name = "lists"]
+struct NewList<'a> {
+    pub name: &'a str,
+    pub user_id: Option<i32>,
+    type_: i32,
+}
+
+#[derive(Clone, Queryable, Identifiable)]
+struct ListElem {
+    pub id: i32,
+    pub list_id: i32,
+    pub user_id: Option<i32>,
+    pub blog_id: Option<i32>,
+    pub word: Option<String>,
+}
+
+#[derive(Default, Insertable)]
+#[table_name = "list_elems"]
+struct NewListElem<'a> {
+    pub list_id: i32,
+    pub user_id: Option<i32>,
+    pub blog_id: Option<i32>,
+    pub word: Option<&'a str>,
+}
+
+impl List {
+    last!(lists);
+    get!(lists);
+
+    fn insert(conn: &Connection, val: NewList) -> Result<Self> {
+        diesel::insert_into(lists::table)
+            .values(val)
+            .execute(conn)?;
+        List::last(conn)
+    }
+
+    pub fn list_for_user(conn: &Connection, user_id: Option<i32>) -> Result<Vec<Self>> {
+        if let Some(user_id) = user_id {
+            lists::table
+                .filter(lists::user_id.eq(user_id))
+                .load::<Self>(conn)
+                .map_err(Error::from)
+        } else {
+            lists::table
+                .filter(lists::user_id.is_null())
+                .load::<Self>(conn)
+                .map_err(Error::from)
+        }
+    }
+
+    pub fn find_by_name(conn: &Connection, user_id: Option<i32>, name: &str) -> Result<Self> {
```

Comment:
if we're searching by name, why do we need the user_id (i understand it's `Option`al, but still)

Reply:
I's a find by (list) name, in context of the given user (or instance if None)

Reply:
that needs either documentation or different names, because this way it's just way too confusing